PAWS/ECGRecordings/ECGModule.swift

```diff
@@ -96,4 +96,133 @@ class ECGModule: Module, DefaultInitializable, EnvironmentAccessible {
+            try await mockWebService.upload(path: "healthkit/\(sample.uuid.uuidString)", body: jsonRepresentation)
+        } else {
+            try await standard.healthKitDocument(id: sample.id).setData(from: resource)
+        }
+    }
+    
+    func upload(electrocardiogram: HKElectrocardiogram) async {
+        var supplementalMetrics: [HKSample] = []
+        
+        do {
+            try await upload(sample: electrocardiogram)
+            
+            supplementalMetrics.append(contentsOf: (try? await electrocardiogram.precedingPulseRates) ?? [])
+            supplementalMetrics.append(contentsOf: (try? await electrocardiogram.precedingPhysicalEffort) ?? [])
+            supplementalMetrics.append(contentsOf: (try? await electrocardiogram.precedingStepCount) ?? [])
+            supplementalMetrics.append(contentsOf: (try? await electrocardiogram.precedingActiveEnergy) ?? [])
+            
+            if let precedingVo2Max = try? await electrocardiogram.precedingVo2Max {
+                supplementalMetrics.append(precedingVo2Max)
+            }
+            
+            for supplementalMetric in supplementalMetrics {
+                do {
+                    try await upload(sample: supplementalMetric)
+                } catch {
+                    logger.log("Could not upload \(supplementalMetric.sampleType): \(error)")
+                    let content = UNMutableNotificationContent()
+                    content.title = "Upload Error"
+                    content.body = "Sample could not be uploaded \(supplementalMetric.sampleType.description) (\(supplementalMetric.uuid.uuidString) at \(Date.now.formatted(date: .numeric, time: .complete)): \((supplementalMetric as? HKQuantitySample)?.quantity.description ?? "Unknown")"
+                    let request = UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: nil)
+                    try? await UNUserNotificationCenter.current().add(request)
+                }
+            }
+        } catch {
+            logger.log("Could not access HealthKit sample: \(error)")
+            await addECGMessage(for: electrocardiogram)
+        }
+    }
+    
+    
+    /// Creates a notification with a title and body message when there is an error accessing a HealthKit sample.
+    /// - Parameter electrocardiogram: The `HKElectrocardiogram` object for which the error occurred.
+    func addECGMessage(for electrocardiogram: HKElectrocardiogram) async {
+        let content = UNMutableNotificationContent()
+        content.title = "HealthKit Error"
+        content.body = "Sample \(electrocardiogram.sampleType.description) with identifier \(electrocardiogram.uuid.uuidString)"
+        let request = UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: nil)
+        try? await UNUserNotificationCenter.current().add(request)
+    }
+    
+    private func uploadUnuploadedECGs() async throws {
```

Comment:
Should this be part of the refresh function or called at the end of the refresh function?